netty/src/main/java/io/grpc/transport/netty/NettyClientTransport.java

```diff
@@ -159,23 +175,27 @@ public void start(Listener transportListener) {
     channel.closeFuture().addListener(new ChannelFutureListener() {
       @Override
       public void operationComplete(ChannelFuture future) throws Exception {
-        if (!future.isSuccess()) {
-          // The close failed. Just notify that transport shutdown failed.
-          notifyTerminated(future.cause());
-          return;
-        }
-
-        if (handler.connectionError() != null) {
-          // The handler encountered a connection error.
-          notifyTerminated(handler.connectionError());
-        } else {
-          // Normal termination of the connection.
-          notifyTerminated(null);
-        }
+        notifyClosed(future);
       }
     });
   }
 
+  /**
+   * Called after the channel has been closed to notify the application that the transport
+   * has been terminated.
+   */
+  private void notifyClosed(ChannelFuture closeFuture) {
+    if (!closeFuture.isSuccess()) {
+      // The close failed, just use the result of the future.
+      failureCause = closeFuture.cause();
```

Comment:
This can overwrite the failureCause; it seems we would want to check it for `null` since the handshake could have failed.